src/account_id_ref.rs

```diff
@@ -29,6 +29,28 @@ use crate::{AccountId, ParseAccountError};
 #[cfg_attr(feature = "abi", derive(schemars::JsonSchema, BorshSchema))]
 pub struct AccountIdRef(pub(crate) str);
 
+/// Enum representing possible types of accounts.
+#[derive(PartialEq)]
+pub enum AccountType {
+    /// Any valid account, that is neither NEAR-implicit nor ETH-implicit.
+    NamedAccount,
+    /// An account with 64 characters long hexadecimal address.
+    NearImplicitAccount,
+    /// An account which address starts with '0x', followed by 40 hex characters.
+    EthImplicitAccount,
+}
+
+impl AccountType {
+    pub fn is_implicit(&self) -> bool {
+        match &self {
+            Self::NearImplicitAccount => true,
+            // TODO(eth-implicit) change to true later, see https://github.com/near/nearcore/issues/10018
+            Self::EthImplicitAccount => false,
+            Self::NamedAccount => false,
+        }
+    }
+}
```

Comment:
What is the motivation for this extra enum? I feel that `is_eth_implicit` and `is_near_implicit` + `is_implicit` would be enough.

Reply:
@frol please see the links in the pr description for more details about this project

The reason for using an enum here is good software development practice. We want to be explicit about the different types of accounts and force ourselves and future developers to be aware and correctly handle the different types of account. Additionally this is more future proof if we ever need to add more account types. This was also discussed on the original PR in nearcore.

Reply:
> We want to be explicit about the different types of accounts and force ourselves and future developers to be aware and correctly handle the different types of account.

If that is the intent, we should remove the `is_near_implicit` and `is_eth_implicit` methods from the AccountIdRef type as having two different ways to do the same thing is not helping here and does not really force users to handle all the cases.

Just for the future reference, I will link the prior discussion: https://github.com/near/nearcore/pull/9969#discussion_r1367162654

Reply:
P.S. Given that "renaming" `is_implicit` to `is_near_implicit` is already a breaking change, removing all those `is*implicit` methods might be just fine.

Just to model the usage here:

BEFORE

```rust
fn send_near(receiver_id: AccountIdRef) {
    if !receiver_id.is_implicit() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

AFTER renaming to `is_near_implicit` and naive changes:

```rust
fn send_near(receiver_id: AccountIdRef) {
    if !receiver_id.is_near_implicit() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

AFTER if no `is_*implicit` methods:

```rust
fn send_near(receiver_id: AccountIdRef) {
    if let AccountType::NamedAccount = receiver_id.get_account_type() {
        assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
    }
    ...
}
```

OR

```rust
fn send_near(receiver_id: AccountIdRef) {
    match receiver_id.get_account_type() {
        AccountType::NamedAccount => {
            assert!(is_account_recorded_on_chain(receiver_id), "you cannot send tokens to non-existing non-implicit accounts");
        }
        AccountType::NearImplicitAccount => {},
        AccountType::EthImplicitAccount => {},
    }
    ...
}
```

Reply:
@frol Would it be ok if I just rename `pub fn is_eth_implicit` to `fn is_eth_implicit` and `pub fn is_near_implicit` to `fn is_near_implicit`? I would remove docstrings with tests from `is_eth_implicit` and `is_near_implicit`, and add corresponding docstring with tests for `get_account_type`.

Reply:
@staffik Do you mean removing `pub` from `is_eth_implicit` and `is_near_implicit`? I would consider moving them to `validation.rs` file

Reply:
@frol Yes. Sure, I will try to move them to `validation.rs` file.